### PR TITLE
🎉🎉🎉 Chakra Core

### PR DESCRIPTION
**Pull request by yufeih**
azure-docs-pr build time:

- Chakra Core: 1:20
- JINT v2: 2:00
- JINT v3: 1:30

https://dev.azure.com/ceapex/Engineering/_workitems/edit/113226/

-----
We gain almost 40 seconds running in CI server:

Before `Build done in 00:01:52.5610000`

After `Build done in 00:01:14.6590000`

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/4995)

### CODE REVIEW

File: azure-pipelines-v3.yml [hunk 1]
```diff
@@ -76,6 +99,13 @@ jobs:
     clean: true
     path: docfx
 
+  # Install .NET Core sdk
+  - task: UseDotNet@2
+    displayName: 'Install .NET Core sdk 2.2.300'
+    inputs:
+      packageType: sdk
+      version: 2.2.401
```
Comment (yufeih) [inline azure-pipelines-v3.yml]: We need to decouple .net core version between docfx and the one installed on the server...

**Comment by lahma** [comment]
Can you compare against Jint 3 beta? It's way faster and has more ES6 support.  /cc @sebastienros

**Comment by yufeih** [comment]
I'll definitely try out Jint 3 tomorrow 👍
________________________________

**Comment by yufeih** [comment]
@lahma @sebastienros I tried JINT 3 today and it is indeed faster then JINT 2, cut javascript execution time by 30s for our main perf target. It is still a little bit slower than ChakraCore but you guys did makes it way faster.

It however throws on some of our test repos, I guess it might be related to parallel execution because it only occurs in Release mode. Callback at the end.

Given that ChakraCore is also cross platform, has good C# interoperability, battle tested by real critical production systems like Edge and CosmosDB, maintained by a dedicated team, I am currently more leaning towards that option.

```
System.ArgumentException: Unable to extract correct key, node type: FunctionExpression
   at Jint.Runtime.ExceptionHelper.ThrowArgumentException(String message, String paramName) in C:\projects\jint\Jint\Runtime\ExceptionHelper.cs:line 35
   at Jint.EsprimaExtensions.GetKey[T](T expression, Engine engine) in C:\projects\jint\Jint\EsprimaExtensions.cs:line 49
   at Jint.Runtime.JavaScriptException.SetCallstack(Engine engine, Nullable`1 location) in C:\projects\jint\Jint\Runtime\JavaScriptException.cs:line 55
   at Jint.Runtime.ExceptionHelper.ThrowJavaScriptException(Engine engine, JsValue value, Completion result) in C:\projects\jint\Jint\Runtime\ExceptionHelper.cs:line 142
   at Jint.Native.Function.ScriptFunctionInstance.Call(JsValue thisArg, JsValue[] arguments) in C:\projects\jint\Jint\Native\Function\ScriptFunctionInstance.cs:line 113
   at Jint.Runtime.Interpreter.Expressions.JintCallExpression.EvaluateInternal() in C:\projects\jint\Jint\Runtime\Interpreter\Expressions\JintCallExpression.cs:line 165
   at Jint.Runtime.Interpreter.Expressions.JintExpression.GetValue() in C:\projects\jint\Jint\Runtime\Interpreter\Expressions\JintExpression.cs:line 32
   at Jint.Runtime.Interpreter.Statements.JintExpressionStatement.ExecuteInternal() in C:\projects\jint\Jint\Runtime\Interpreter\Statements\JintExpressionStatement.cs:line 18
   at Jint.Runtime.Interpreter.JintStatementList.Execute() in C:\projects\jint\Jint\Runtime\Interpreter\JintStatementList.cs:line 69
   at Jint.Native.Function.ScriptFunctionInstance.Call(JsValue thisArg, JsValue[] arguments) in C:\projects\jint\Jint\Native\Function\ScriptFunctionInstance.cs:line 67
   at Jint.Native.JsValue.Invoke(JsValue thisObj, JsValue[] arguments) in C:\projects\jint\Jint\Native\JsValue.cs:line 399
   at Microsoft.Docs.Build.JavascriptEngine.Run(String scriptPath, String methodName, JToken arg) in D:\docfx\src\docfx\template\JavascriptEngine.cs:line 54
   at Microsoft.Docs.Build.TemplateEngine.RunJint(String scriptName, JObject model, String methodName, Boolean tryParseFromContent) in D:\docfx\src\docfx\template\TemplateEngine.cs:line 98
   at Microsoft.Docs.Build.BuildTableOfContents.Build(Context context, Document file) in D:\docfx\src\docfx\build\toc\BuildTableOfContents.cs:line 45
   at Microsoft.Docs.Build.Build.BuildFile(Context context, Document file) in D:\docfx\src\docfx\build\Build.cs:line 121
   at Microsoft.Docs.Build.Build.Run(Docset docset, Docset fallbackDocset, RestoreGitMap restoreGitMap, CommandLineOptions options, ErrorLog errorLog, String outputPath) in D:\docfx\src\docfx\build\Build.cs:line 68
   at Microsoft.Docs.Build.Build.Run(String docsetPath, CommandLineOptions options, ErrorLog errorLog) in D:\docfx\src\docfx\build\Build.cs:line 36
   at Microsoft.Docs.Build.Docfx.Run(String[] args) in D:\docfx\src\docfx\cli\Docfx.cs:line 90
   at Microsoft.Docs.Build.Docfx.Main(String[] args) in D:\docfx\src\docfx\cli\Docfx.cs:line 25
```

**Comment by sebastienros** [comment]
The engine is not thread safe. And I don't think any JS engine is. You should have one instance per thread if you are processing templates in parallel.

I assume Jint makes more sense when you need to have a more secure environment. For instance RavenDB uses it because you can:
- limit the number of statements, to prevent infinite loops
- limit the recursion level, to prevent infinite recursion
- limit the memory usage
- prevent unauthorized access to .NET types

If you don't need these and Chakra suits your other needs then that's reasonable.

**Comment by lahma** [comment]
> System.ArgumentException: Unable to extract correct key, node type: FunctionExpression

@yufeih would it be possible for you to share the lines of code that trigger this behavior? Shouldn't be too hard to fix unless you rely on some ES6 feature that hasn't been completed fully yet.

**Comment by yufeih** [comment]
> > System.ArgumentException: Unable to extract correct key, node type: FunctionExpression
> 
> @yufeih would it be possible for you to share the lines of code that trigger this behavior? Shouldn't be too hard to fix unless you rely on some ES6 feature that hasn't been completed fully yet.

@lahma , I suspect it is caused by my misuse of parallel, the same engine is shared across multiple thread.

File: src/docfx/lib/js/ChakraCoreJsEngine.cs [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+
+                    return ToJToken(output);
+                }
+                catch (JavaScriptScriptException ex) when (ex.Error.IsValid)
+                {
+                    throw new JavaScriptEngineException($"{ex.ErrorCode}:\n{ToJToken(ex.Error)}");
+                }
+                finally
+                {
+                    Native.ThrowIfError(Native.JsSetCurrentContext(JavaScriptContext.Invalid));
```
Comment (mingwli) [inline src/docfx/lib/js/ChakraCoreJsEngine.cs]: >Native.ThrowIfError(Native.JsSetCurrentContext(JavaScriptContext.Invalid)) [](start = 20, length = 74)

if current context is invalid, why do we add it to the context pool?

File: src/docfx/lib/js/ChakraCoreJsEngine.cs [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+
+        private static readonly JavaScriptNativeFunction s_requireFunction = new JavaScriptNativeFunction(Require);
+
+        private static readonly ThreadLocal<Stack<string>> t_dirnames = new ThreadLocal<Stack<string>>(() => new Stack<string>());
+
+        private static readonly ThreadLocal<Dictionary<string, JavaScriptValue>> t_modules = new ThreadLocal<Dictionary<string, JavaScriptValue>>();
+
+        private static JavaScriptSourceContext s_currentSourceContext = JavaScriptSourceContext.FromIntPtr(IntPtr.Zero);
+
+        private readonly ConcurrentDictionary<JavaScriptContext, JavaScriptValue> _globals = new ConcurrentDictionary<JavaScriptContext, JavaScriptValue>();
```
Comment (live1206) [inline src/docfx/lib/js/ChakraCoreJsEngine.cs]: should this be `ConcurrentDictionary<JavaScriptContext, Lazy<JavaScriptValue>>`? same as `s_scriptExports`

File: src/docfx/template/TemplateEngine.cs [hunk 4]
```diff
@@ -33,7 +34,13 @@ private TemplateEngine(string templateDir)
             _global = LoadGlobalTokens();
             _schemas = LoadSchemas(schemaDir, _contentTemplateDir);
             _liquid = new LiquidTemplate(_contentTemplateDir);
-            _js = new JavascriptEngine(_contentTemplateDir, _global);
+
+            // TODO: remove JINT after Microsoft.CharkraCore NuGet package
+            // supports linux and macOS: https://github.com/microsoft/ChakraCore/issues/2578
+            _js = RuntimeInformation.IsOSPlatform(OSPlatform.Windows)
+                ? (IJavaScriptEngine)new ChakraCoreJsEngine(_contentTemplateDir, _global)
```
Comment (live1206) [inline src/docfx/template/TemplateEngine.cs]: why cast to `IJavaScriptEngine` here? _js is already `IJavaScriptEngine`

File: src/docfx/lib/js/ChakraCoreJsEngine.cs (hunk 3 above)
Comment (yufeih) [reply src/docfx/lib/js/ChakraCoreJsEngine.cs]: evaluation of  `s_scriptExports` is relatively expensive, requires the script to run to end, so I put a lazy there, but the evaluation of `_globals` is very fast, just marshaling data.

File: src/docfx/template/TemplateEngine.cs (hunk 4 above)
Comment (yufeih) [reply src/docfx/template/TemplateEngine.cs]: C# is not smart enough to deduce the type here

**Comment by sebastienros** [comment]
Is there some documentation somewhere explaining how to run the integration tests? I tried cloning this repos and the azure docs one but got lots of warnings. Am i doing something wrong?

**Comment by yufeih** [comment]
@sebastienros unfortunately it is currently not possible to build azure docs using only docfx v3 at the moment, The javascript we execute during integration test comes from docs template and that part is still private.
